### PR TITLE
Make Psi4 buildable with Ninja

### PR DESCRIPTION
**Pull request by robertodr**
## Description
Ninja is supposed to be faster at compiling than make. CMake supports it out-of-the-box, it's enough to pass it as generator:
```
cmake -H. -Bbuild -GNinja
```
I have tried to use Ninja on Travis and the Clang build runs in ~21 minutes. The GCC builds however fail due to a compiler bug.

## Todos
Notable points that this PR has either accomplished or will accomplish.
* **Developer Interest**
  - [x] `DESTDIR=${CMAKE_BINARY_DIR}/stage` was prepended to `${CMAKE_MAKE_PROGRAM} install` appearing in all instances of `INSTALL_COMMAND`.
  - [x] Same `DESTDIR`-related modifications in Libint and update of the Git commit SHA. PR towards v1 was already merged. 
* **User-Facing for Release Notes**
  - [x] Psi4 core now buildable with Ninja. **Caveat** To build also the addons it might be necessary to use CMake v3.7.2 _and_ the [Kitware-maintained version of Ninja](https://github.com/Kitware/ninja/releases). The official one does not support Fortran.

## Status
- [x] Ready to go

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -139,6 +134,10 @@ before_install:
   - df -h
   - ulimit -a
 install:
+- export NINJA_URL="https://github.com/Kitware/ninja/releases/download/v1.7.2.gaad58.kitware.dyndep-1/"
+- export NINJA_TARBALL_NAME="ninja-1.7.2.gaad58.kitware.dyndep-1_x86_64-linux-gnu"
+- curl -L ${NINJA_URL}${NINJA_TARBALL_NAME}.tar.gz | tar xz
+- export PATH=$PWD/${NINJA_TARBALL_NAME}:$PATH
```
Comment (loriab) [inline .travis.yml]: looks like you can `conda install ninja -c conda-forge` for 1.7.2

Comment (robertodr) [reply .travis.yml]: True, thanks for pointing that out. Unfortunately they ship the official v1.7.2, while only the Kitware-maintained fork supports Fortran.

File: CMakeLists.txt [hunk 2]
```diff
@@ -223,8 +223,7 @@ ExternalProject_Add(psi4-core
               -DDESTDIR=${CMAKE_BINARY_DIR}/stage
    CMAKE_CACHE_ARGS -DCMAKE_PREFIX_PATH:PATH=${CMAKE_PREFIX_PATH}
    BUILD_ALWAYS 1
-   INSTALL_COMMAND ${CMAKE_MAKE_PROGRAM} install
-                   DESTDIR=${CMAKE_BINARY_DIR}/stage)
+   INSTALL_COMMAND DESTDIR=${CMAKE_BINARY_DIR}/stage ${CMAKE_MAKE_PROGRAM} install)
```
Comment (loriab) [inline CMakeLists.txt]: Oh my, lots of upstream PRs to get this change percolated :-)

File: external/upstream/libint/CMakeLists.txt [hunk 3]
```diff
@@ -9,7 +9,7 @@ else()
     message(STATUS "Suitable Libint could not be located, ${Magenta}Building Libint${ColourReset} instead.")
     ExternalProject_Add(libint_external
         GIT_REPOSITORY https://github.com/evaleev/libint
-        GIT_TAG 3f14ff0  # v1.2.1 release-1-2-1
+        GIT_TAG 024738cf # v1.2.1 release-1-2-1
```
Comment (loriab) [inline external/upstream/libint/CMakeLists.txt]: Thanks for handling this w/Libint.

**Comment by robertodr** [comment]
@loriab yes, that still works. Travis still uses Makefiles, since the GCC builds are failing with a compiler error with Ninja.

**Comment by dgasmith** [comment]
Just for fun, can you run a timing comparison with and without Ninja?

**Comment by robertodr** [comment]
Not a huge success @dgasmith...
The Clang build is crashing for unclear reasons:
https://travis-ci.org/robertodr/psi4/jobs/280959304
Compiler chokes on some `export_*.cc` for the GCC builds: 
https://travis-ci.org/robertodr/psi4/jobs/280959306
https://travis-ci.org/robertodr/psi4/jobs/280959309

**Comment by robertodr** [comment]
Setup line:
```
cmake -H. -Bbuild 
            -DCMAKE_CXX_COMPILER=icpc 
            -DCMAKE_Fortran_COMPILER=ifort 
            -DCMAKE_C_COMPILER=icc
            -DENABLE_dkh=ON
```
Compiling with:
```
cmake --build . -- -j20
```
Results for `-GNinja`:
```
real	13m45.668s
user	80m17.346s
sys	10m27.013s
```
Results for `-G"Unix Makefiles"`:
```
real	13m14.431s
user	79m20.198s
sys	10m9.984s
```

**Comment by dgasmith** [comment]
Hmm, do you think we would see anything else for another compiler like `clang`? Might be good to document this, get it in, and have it be disabled incase Ninja can find more optimizations in the future.

**Comment by robertodr** [comment]
It is entirely possible that it works better with `clang`. I'll try to run the same test.

**Comment by robertodr** [comment]
Setup line:
```
cmake -H. -Bbuild 
            -DCMAKE_CXX_COMPILER=clang++ 
            -DCMAKE_CXX_FLAGS="--gcc-toolchain=/global/hds/software/cpu/eb3/GCCcore/5.4.0"
            -DCMAKE_Fortran_COMPILER=gfortran
            -DCMAKE_C_COMPILER=clang
            -DCMAKE_C_FLAGS="--gcc-toolchain=/global/hds/software/cpu/eb3/GCCcore/5.4.0"
            -DPYTHON_EXECUTABLE=`type -P python`
            -DENABLE_dkh=ON
```
Compiling with:
```
cmake --build . -- -j20
```
Results for `-GNinja`:
```
real	41m37.081s
user	458m12.064s
sys	2m24.967s
```
Results for `-G"Unix Makefiles"`:
```
real	45m45.169s
user	458m39.363s
sys	2m28.485s
```
I build Clang in debug mode, I guess that's why the builds are overall slower than with the Intel compilers. I am recompiling Clang in release mode to check that.

**Comment by robertodr** [comment]
OK, with Clang built in release mode I get:
- For `-G"Ninja"`
```
real	7m35.581s
user	31m11.930s
sys	2m7.600s
```
- For `-G"Unix Makefiles"`
```
real	8m9.513s
user	31m30.395s
sys	3m24.407s
```

**Comment by dgasmith** [comment]
@robertodr Just to double check, `make -j20` will still work fine. This just allows the optional to build with Ninja?

**Comment by loriab** [comment]
Yeah, I can't tell which generator travis is using just from the build log. Because of the Ninja+Fortran+CMake issues, I don't think we can have Ninja as the default cmake generator. But if this is just enabling Ninja as an available non-default generator, LGTM.

**Comment by robertodr** [comment]
@loriab @dgasmith Everything is as before on Travis. This only makes the option to build with Ninja work with Psi4 if the constraints on CMake version on Ninja version are satisfied.

**Pull request merged**